### PR TITLE
Build: Add notice file generation

### PR DESCRIPTION
**Pull request by rjernst**
This change improves the notice file present in our distributions to
include notice and license files from each included dependency.

closes #22546

### CODE REVIEW

File: distribution/build.gradle [hunk 1]
```diff
@@ -40,6 +41,20 @@ buildscript {
...
+ *****************************************************************************/
+
+// integ test zip only uses core, so a different notice file is needed there
+task buildCoreNotice(type: NoticeTask) {
+  dependencies project(':core')
+}
+
+// other distributions include notices from modules as well, which are added below later
+task buildFullNotice(type: NoticeTask) {
+  dependencies project(':core')
```
Comment (nik9000) [inline distribution/build.gradle]: This is the same as the one above.

File: distribution/build.gradle [hunk 2]
```diff
@@ -172,10 +190,17 @@ subprojects {
...
     commonFiles = copySpec {
       from rootProject.projectDir
-      include 'NOTICE.txt'
       include 'LICENSE.txt'
       include 'README.textile'
     }
+
+    noticeFile = copySpec {
+      if (project.name == 'integ-test-zip') {
+        from buildCoreNotice
```
Comment (nik9000) [inline distribution/build.gradle]: I can't find this name. Does that resolve?

File: distribution/build.gradle (hunk 1 above)
Comment (nik9000) [inline distribution/build.gradle]: Can you document why we need both? Or maybe we don't need both and this is a leftover?

Comment (rjernst) [inline distribution/build.gradle]: I tried explaining it in the comment above, but I guess it was too terse.  Integ test zip does not have modules, so I generate a notice just with deps from core for that, while the others have the "full" notice file, which includes both core, and modules dependencies.

File: distribution/build.gradle (hunk 2 above)
Comment (rjernst) [inline distribution/build.gradle]: It is defined on line 48.

Comment (nik9000) [inline distribution/build.gradle]: Ah. Sorry. Sometimes I read those as function definitions....

**Review by nik9000 — APPROVED** [review]

**Pull request merged**